sample/samplers.go

```diff
@@ -2,59 +2,75 @@ package sample
 
 import (
 	"errors"
-	"math"
-
-	"golang.org/x/exp/rand"
-	"gonum.org/v1/gonum/stat/sampleuv"
+	"math/rand/v2"
 )
 
+// Sampler is not thread-safe. Each goroutine should have its own instance.
 type Sampler interface {
 	Sample([]float32) (int32, error)
 }
 
+type tokenInfo struct {
+	id    int32
+	logit float32
+	prob  float32
+}
+
+type tokenSliceInfo struct {
+	tokens []tokenInfo
+	sorted bool
+	sum    float32
+}
+
 type weighted struct {
-	src        rand.Source
-	transforms []Transform
+	rng        *rand.Rand
+	transforms []transform
 }
 
-// TODO(parthsareen): remove uv sample dependency https://github.com/ollama/ollama/issues/9279
-func Weighted(seed *uint64, transforms ...Transform) Sampler {
-	var src rand.Source
-	if seed != nil {
-		src = rand.NewSource(*seed)
+func Weighted(rng *rand.Rand, transforms ...transform) Sampler {
+	return &weighted{
+		rng:        rng,
+		transforms: transforms,
 	}
-	return weighted{src: src, transforms: transforms}
 }
 
-func (s weighted) Sample(logits []float32) (int32, error) {
-	logits64 := make([]float64, len(logits))
+func (s *weighted) Sample(logits []float32) (int32, error) {
+	tokens := make([]tokenInfo, len(logits))
 	for i, v := range logits {
-		logits64[i] = float64(v)
+		tokens[i].id = int32(i)
+		tokens[i].logit = v
 	}
 
+	tokensInfo := tokenSliceInfo{tokens: tokens, sorted: false}
 	for _, t := range s.transforms {
-		logits64 = t.Apply(logits64)
+		tokensInfo = t.Apply(tokensInfo)
 	}
 
-	logitsCopy := make([]float64, 0, len(logits))
-	indices := make([]int, 0, len(logits))
-	for i, logit := range logits64 {
-		if !math.IsInf(logit, -1) {
-			logitsCopy = append(logitsCopy, logit)
-			indices = append(indices, i)
-		}
+	if len(tokensInfo.tokens) == 0 {
+		return -1, errors.New("no valid logits found for weighted sampling")
 	}
 
-	if len(logitsCopy) == 0 {
-		return -1, errors.New("no valid logits found for weighed sampling")
+	// TODO(parthsareen): fix seeded sampling https://github.com/ollama/ollama/issues/9554
+	var r float32
+	if s.rng != nil {
+		r = s.rng.Float32()
```

Comment:
Can we instead always use an explicit `math/rand`?

Reply:
IIUC for deterministic sampling if a seed is provided we should be using the global one right?